### PR TITLE
[Feature] Support rewrite generated expression to generated column

### LINKED ISSUE #20305: [Feature] Support Materialized Column
Materialized Column is a kind of column that can evaluated through other columns in the table. The expensive computation can be done when loading the data and speed up the query using this technique.

usage:
1. using CREATE TABLE:

CREATE TABLE t ( id BIGINT NOT NULL,  array_int ARRAY<int> NOT NULL, array_string ARRAY<string> NOT NULL, json_data JSON NOT NULL,
mc_1 DOUBLE MATERIALIZED_COLUMN (array_avg(array_int)),
mc_2 ARRAY<int> MATERIALIZED_COLUMN (array_sort(array_int)),
mc_3 ARRAY<string> MATERIALIZED_COLUMN (array_sort(array_string)),
mc_4 String MATERIALIZED_COLUMN (get_json_string(json_string(json_data), "$.a")) )
Primary KEY (id) DISTRIBUTED BY HASH(id) BUCKETS 48 PROPERTIES("replication_num" = "1", "replicated_storage"="true");

2. using ALTER TABLE:
ALTER TABLE t ADD MATERIALIZED COLUMN mc_5 BIGINT (id);

### PR DESCRIPTION
**Pull request by HangyuanLiu**
## What type of PR is this:
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes:

Fixes #20305

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 39 / 39 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/QueryTransformer.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/TableRelation.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/SelectRelation.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzeState.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/ExpressionMapping.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|16|16|100.00%|[]|

**Review by srlch — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 39 / 39 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/QueryTransformer.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/TableRelation.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/SelectRelation.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzeState.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/ExpressionMapping.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|16|16|100.00%|[]|

**Comment by nshangyiming** [comment]
@kangkaisen could you help to review this pr

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/SelectRelation.java [hunk 1]
```diff
@@ -77,6 +79,12 @@ public class SelectRelation extends QueryRelation {
 
     private Map<Expr, FieldId> columnReferences;
 
+    /**
+     *  materializeExpressionToColumnRef stores the mapping relationship
+     *  between materialized expressions and materialized columns
+     */
+    private Map<Expr, SlotRef> materializeExpressionToColumnRef = new HashMap<>();
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/ast/SelectRelation.java:86]: yep. Suggest `materialized` instead of `materialize` ?

**Review by murphyatwork — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 39 / 39 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/QueryTransformer.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/TableRelation.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/SelectRelation.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzeState.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/ExpressionMapping.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|16|16|100.00%|[]|

**Review by murphyatwork — APPROVED** [review]

**Review by LiShuMing — APPROVED** [review]

**Review by Astralidea — APPROVED** [review]

**Pull request merged**